ipatests/test_webui/test_loginscreen.py

```diff
@@ -0,0 +1,408 @@
+        """
+        if (self.logged_in()):
+            self.logout()
+        self.load_url(self.get_base_url())
+        self.init_app(loginscreen.PKEY, loginscreen.PASSWD_ITEST_USER_NEW)
+        # User is not logged in
+        assert self.logged_in()
+
+    def reset_password(self, username=None, current_password=None,
+                       new_password=None, link_text=None, link_url=None):
+        """
+        Reset password with the given one
+        """
+        login_screen = self.get_login_screen()
+
+        if username is not None:
+            username_field = self.get_input_field('username', login_screen)
+        cur_pass_field = self.get_input_field('current_password', login_screen)
+        new_pass_field = self.get_input_field('new_password', login_screen)
+        verify_pass_field = self.get_input_field('verify_password',
+                                                 login_screen)
+        if username is not None:
+            username_field.send_keys(username)
+        cur_pass_field.send_keys(current_password)
+        new_pass_field.send_keys(new_password)
+        verify_pass_field.send_keys(new_password)
+        verify_pass_field.send_keys(Keys.RETURN)
+        self.wait()
+        self.assert_notification(assert_text='Password change complete',
+                                 link_text=link_text, link_url=link_url)
+
+    def get_data_from_form_row(self, form_row):
+        """
+        Parse data from the form record to a comparable structure
+        """
+        result = []
+        label = self.find_xelement(".//label", form_row)
+        assert label.is_displayed()
+        result.append(label.get_attribute('name'))
+        result.append(label.text)
+        req = self.find_xelement("./../..", label)
+        result.append(self.has_class(req, 'required'))
+
+        field = self.find_xelement(".//input", form_row)
+        # not editable field
+        editable = field.is_displayed()
+        if not editable:
+            field = self.find_xelement(".//p", form_row)
+
+        result.append(editable)
+        assert field.is_displayed()
+        result.append(field.get_attribute('type'))
+        result.append(field.get_attribute('name'))
+        if not editable:
+            result.append(field.text)
+        else:
+            result.append(field.get_attribute('value'))
+        result.append(field.get_attribute('placeholder'))
+
+        return tuple(result)
+
+    def get_data_from_button(self, button):
+        """
+        Parse data from the button to a comparable structure
+        """
+        result = []
+        result.append(button.get_attribute('name'))
+        result.append(button.get_attribute('title'))
+
+        return tuple(result)
+
+    def get_data_from_alert(self, data):
+        """
+        Parse data from the alert to a comparable structure
+        """
+        result = []
+        result.append(data.text)
+
+        return tuple(result)
```

Comment:
Ah, may be i see :)
Are you talking about 'get_data_from_alert'?

Reply:
@serg-cymbaluk , this is applied for consistency of the other similar methods.
In future may not only 'data.text' will be gathered.
For now it is redundant.

Reply:
There should be as much lines of code as operations regardless of future changes.
Because code is becoming unreasonably complicated.

Reply:
https://github.com/freeipa/freeipa/pull/2151#discussion_r213557470
It is also some future change ;-)
Anyway, I don't dispute you.

Reply:
Updated and pushed.
Please refer.